File: ts_sdk/src/agentic_processor/agentic-processor.ts
import { APIEntity, dataManager, registerEntity } from '../APIEntity';
import { IEntity } from '../IEntity';
import { ActionInfo } from '../models/ActionInfo';
import { TypeId } from '../models/TypeId';
import { FlowData } from '../flow_processing';
import { AgenticContext, serializeAgenticContext } from './agentic-context';
import { AgenticProcess, IAgenticProcess } from './agentic-process';
import { InstructionFile } from '../models/workflow/InstructionFile';

// Import shared types from agentic-types (breaks circular dependency)
import {
  ProcessorStatus,
  UIComponentPayload,
  ParsedUIUri,
  parseUIUri,
} from './agentic-types';

// Re-export so existing imports from './agentic-processor' continue to work
export { ProcessorStatus, parseUIUri };
export type { UIComponentPayload, ParsedUIUri };

/**
 * Interface for AgenticProcessor entity data
 */
export interface IAgenticProcessor extends IEntity {
  active_agentic_process_id?: string;
}

export interface CreateProcessOptions {
  /** Optional result metadata to create a ProcessResult child */
  result?: {
    uname?: string;
    resultType?: string;
    sourceSessionId?: string;
  };
  /** Whether to watch the processor before creating the process (default: true) */
  watchProcessor?: boolean;
  /** Whether to watch the created process immediately (default: true) */
  watchProcess?: boolean;
  /** Whether the process is visible in the tabs view (default: false) */
  visible?: boolean;
}

/**
 * AgenticProcessor Entity - Backend-synced processor for MDO instruction execution
 *
 * Extends APIEntity to receive entity notifications from the backend.
 * Uses dataManager for entity operations and receives FlowData via handleFlowData.
 *
 * Usage:
 * ```typescript
 * const computeNode = await ComputeNode.getLocal();
 * const processor = await computeNode.createAgenticProcessor();
 *
 * processor.on('ui', (payload) => {
 *   console.log('UI component:', payload);
 * });
 *
 * processor.on('waiting', (inputId) => {
 *   console.log('Waiting for input:', inputId);
 *   processor.sendInput({ answer: 'user response' });
 * });
 *
 * const process = await processor.run(instructionFile, context);
 * for await (const flowData of process.output()) {
 *   console.log('Output:', flowData);
 * }
 * ```
 */
@registerEntity
export class AgenticProcessor extends APIEntity<AgenticProcessor> implements IAgenticProcessor {
  /** Entity type for AgenticProcessor */
  static type: string = 'agentic_processor';

  /** Active agentic process ID if running */
  active_agentic_process_id?: string;

  /** Running processes map */
  private _runningProcesses: Map<string, AgenticProcess> = new Map();

  /** Collected FlowData */
  private _receivedFlowData: FlowData[] = [];

  constructor(entity: Partial<IAgenticProcessor> = {}) {
    // Backward compatibility: API tests still create "apu-{id}" TypeIds and
    // pass them into AgenticProcessor. Normalize legacy alias to this entity.
    const normalizedEntity = { ...(entity as any) };
    if (normalizedEntity.type === 'apu') {
      delete normalizedEntity.type;
    }

    super(normalizedEntity);
  }

  /**
   * Backward-compatible alias used by older APU tests.
   */
  get apuTypeId(): TypeId {
    return new TypeId('apu', this.id);
  }

  /**
   * Handle incoming FlowData from backend entity notification.
   * Overrides APIEntity.handleFlowData to also track FlowData
   * and emit typed events.
   */
  handleFlowData(flowData: FlowData): void {
    // Some backends emit duplicate FlowData for legacy processor/process wiring.
    // Ignore immediate duplicates so UI handlers stay deterministic in tests.
    const lastFlowData = this._receivedFlowData[this._receivedFlowData.length - 1];
    if (lastFlowData && this._isDuplicateFlowData(lastFlowData, flowData)) {
      return;
    }

    // Call parent to ingest to stream (with consolidation) and emit base event
    super.handleFlowData(flowData);

    this._receivedFlowData.push(flowData);

    // Do not forward FlowData to processes here.
    // The backend already emits flow_data to both processor and process entities,
    // and forwarding causes duplicate streaming in the process UI.

    // Check for completion marker - close ALL open groups
    if (flowData.attributes?.['complete'] === 'true') {
      this.flowDataStream.closeOpenGroups();
    }

    // Check for UI element
    const elementType = flowData.attributes?.['element-type'];
    if (elementType === 'ui') {
      const uiPayload = flowData.data as UIComponentPayload;
      this.emit('ui', uiPayload);
    }
  }

  /**
   * Called when entity data is updated via WebSocket notification.
   * Updates local state and emits events.
   */
  protected onEntityUpdate(_data: Partial<IAgenticProcessor>): void {
    // Status changes are pushed via AgenticProcess entity updates, not processor updates.
  }

  /**
   * Run instruction content or file with optional context.
   *
   * This is the unified interface for executing AMD instructions.
   * Returns an AgenticProcess that can be used to stream outputs and track state.
   *
   * @param input - InstructionFile, VFS path string, or raw MDO content string
   * @param contextOrOptions - AgenticContext (if input is InstructionFile) or options
   * @param options - Optional debug settings
   * @returns AgenticProcess for streaming outputs and state tracking
   *
   * @example
   * ```typescript
   * // With InstructionFile and context
   * const instructionFile = InstructionFile.fromContent(amdCode);
   * const context = { computeNode: await getLocalComputeNode() };
   * const process = await processor.run(instructionFile, context);
   *
   * // With VFS path
   * const process = await processor.run('/path/to/skill.md');
   *
   * // With raw content
   * const process = await processor.run('<!-- <flow-do> -->Say hello<!-- </flow-do> -->');
   *
   * for await (const flowData of process.output()) {
   *   console.log('Output:', flowData);
   * }
   * ```
   */
  public async run(
    input: InstructionFile | string,
    contextOrOptions?: AgenticContext | { sourceVfsPath?: string; debug?: boolean; breakpoints?: string[] },
    options?: { debug?: boolean; breakpoints?: string[] },
  ): Promise<AgenticProcess> {
    await this.watch();

    let actionName: string;
    let bodyParams: Record<string, unknown>;
    let context: AgenticContext | undefined;
    let instructionFile: InstructionFile | undefined;

    if (input instanceof InstructionFile) {
      context = contextOrOptions as AgenticContext;
      instructionFile = input;
      actionName = 'run';
      bodyParams = {
        instruction_content: input.content,
        context: context ? serializeAgenticContext(context) : {},
      };
    } else if (typeof input === 'string') {
      const entityTypePattern = /^[a-z_]+-[a-f0-9-]+\//i;
      const isVfsPath =
        input.startsWith('/') ||
        input.startsWith('vfs://') ||
        input.includes('/.flow/') ||
        entityTypePattern.test(input);

      if (isVfsPath) {
        actionName = 'runFile';
        bodyParams = { vfs_path: input };
      } else {
        actionName = 'run';
        bodyParams = { instruction_content: input };
      }
    } else {
      throw new Error('Invalid input: expected InstructionFile or string');
    }

    const actionInfo = new ActionInfo(actionName, AgenticProcessor.type, this.id, 'POST');
    actionInfo.bodyParameters = bodyParams;

    const response = await dataManager.callAction<unknown, IAgenticProcess>(actionInfo);
    const process = dataManager.updateEntityFromJson<AgenticProcess>(response);

    if (instructionFile) {
      process._instructionFile = instructionFile;
    }
    if (context) {
      process._context = context;
    }

    this._runningProcesses.set(process.id, process);
    await process.watch();

    process.on('complete', () => { this._runningProcesses.delete(process.id); });
    process.on('error', () => { this._runningProcesses.delete(process.id); });

    return process;
  }

  /**
   * @deprecated Use run() instead. Will be removed in future version.
   */
  public async start(
    mdoContent: string,
    options?: { sourceVfsPath?: string; debug?: boolean; breakpoints?: string[] },
  ): Promise<AgenticProcess> {
    return this.run(mdoContent, options);
  }

  /**
   * @deprecated Use run() instead. Will be removed in future version.
   */
  public async runFile(
    vfsPath: string,
    options?: { debug?: boolean; breakpoints?: string[] },
  ): Promise<AgenticProcess> {
    return this.run(vfsPath, options);
  }

  /**
   * Execute instruction content directly (simplified API).
   *
   * This is a simpler alternative to run() that takes instruction text
   * directly. Uses initialize_from_prompt() on the backend for single-instruction
   * execution.
   *
   * @param instructionContent - Plain text or AMD instruction content
   * @param context - AgenticContext with compute node and settings
   * @param options - Optional debug settings
   * @returns AgenticProcess for streaming outputs and state tracking
   *
   * @example
   * ```typescript
   * const computeNode = await ComputeNode.getLocal();
   * const processor = await computeNode.createAgenticProcessor();
   * const context = { workdir: '/path/to/project' };
   *
   * // Execute a simple instruction
   * const process = await processor.execute("List all Python files", context);
   * for await (const flowData of process.output()) {
   *   console.log('Output:', flowData);
   * }
   *
   * // Inject additional instructions during execution
   * await process.inject("Now count them");
   * ```
   */
  public async execute(
    instructionContent: string,
    context: AgenticContext,
    options?: { debug?: boolean; breakpoints?: string[] },
  ): Promise<AgenticProcess> {
    await this.watch();

    const actionInfo = new ActionInfo('execute', AgenticProcessor.type, this.id, 'POST');
    actionInfo.bodyParameters = {
      instruction_content: instructionContent,
      context: serializeAgenticContext(context),
      debug: options?.debug ?? false,
      breakpoints: options?.breakpoints ?? [],
    };

    // Backend creates process and returns entity data
    const response = await dataManager.callAction<unknown, IAgenticProcess>(actionInfo);

    // Reuse cached entity when available to avoid duplicate instances/listeners
    const process = dataManager.updateEntityFromJson<AgenticProcess>(response);

    // Store local references
    process._context = context;

    this._runningProcesses.set(process.id, process);

    // Watch the process entity for updates
    await process.watch();

    // Optimistically echo user message into the stream to avoid missing it before watch is active
    process.appendUserMessage(instructionContent);

    // Cleanup when process completes
    process.on('complete', () => {
      this._runningProcesses.delete(process.id);
    });

    process.on('error', () => {
      this._runningProcesses.delete(process.id);
    });

    return process;
  }

  /**
   * Create an idle process ready for execute() calls.
   *
   * This creates a process in IDLE status that can accept instructions
   * via process.executeInstruction(). The process stays alive until
   * explicitly terminated via process.exit().
   *
   * @param context - AgenticContext with compute node and settings
   * @returns AgenticProcess in IDLE status ready for instructions
   *
   * @example
   * ```typescript
   * const computeNode = await ComputeNode.getLocal();
   * const processor = await computeNode.createAgenticProcessor();
   * const process = await processor.createProcess(context);
   *
   * // Execute instructions on the process
   * await process.executeInstruction("Remember the number 42");
   * await process.executeInstruction("What's the number?");
   *
   * // Cleanup when done
   * await process.exit();
   * ```
   */
  public async createProcess(context: AgenticContext, options?: CreateProcessOptions): Promise<AgenticProcess> {
    if (options?.watchProcessor !== false) {
      await this.watch();
    }

    const actionInfo = new ActionInfo('createProcess', AgenticProcessor.type, this.id, 'POST');
    actionInfo.bodyParameters = {
      context: serializeAgenticContext(context),
      ...(options?.result
        ? {
            result: {
              uname: options.result.uname,
              resultType: options.result.resultType,
              sourceSessionId: options.result.sourceSessionId,
            },
          }
        : {}),
      ...(options?.visible !== undefined ? { visible: options.visible } : {}),
    };

    // Backend creates process and returns entity data
    const response = await dataManager.callAction<unknown, IAgenticProcess>(actionInfo);

    // Reuse cached entity when available to avoid duplicate instances/listeners
    const process = dataManager.updateEntityFromJson<AgenticProcess>(response);

    // Store local references
    process._context = context;

    this._runningProcesses.set(process.id, process);

    // Watch the process entity for updates
    if (options?.watchProcess !== false) {
      await process.watch();
    }

    return process;
  }

  /**
   * Get all running processes.
   */
  public getRunningProcesses(): Record<string, AgenticProcess> {
    const result: Record<string, AgenticProcess> = {};
    this._runningProcesses.forEach((process, id) => {
      result[id] = process;
    });
    return result;
  }

  /**
   * Get a specific running process by ID.
   */
  public getProcess(agenticProcessId: string): AgenticProcess | undefined {
    return this._runningProcesses.get(agenticProcessId);
  }

  /**
   * Get all received FlowData
   */
  public getReceivedFlowData(): readonly FlowData[] {
    return [...this._receivedFlowData];
  }

  /**
   * Cleanup and dispose processor
   */
  public dispose(): void {
    // Clear running processes
    for (const process of this._runningProcesses.values()) {
      process._markComplete();
    }
    this._runningProcesses.clear();
  }

  private _isDuplicateFlowData(a: FlowData, b: FlowData): boolean {
    const attrsA = a.attributes || {};
    const attrsB = b.attributes || {};

    return (
      attrsA['element-type'] === attrsB['element-type'] &&
      attrsA['ui-id'] === attrsB['ui-id'] &&
      attrsA['i'] === attrsB['i'] &&
      attrsA['t'] === attrsB['t']
    );
  }

}

File: flow_sdk/builtin/agentic_process/agentic_processor.py
"""AgenticProcessor entity — creates and manages AgenticProcess instances.

Supports the AgenticProcessor action surface:
- POST /api/v1/graph/agentic_processor/{id}/runFile
- POST /api/v1/graph/agentic_processor/{id}/run
- POST /api/v1/graph/agentic_processor/{id}/execute
- POST /api/v1/graph/agentic_processor/{id}/createProcess

Status is transcript-derived: call GET /api/v1/graph/agentic_process/{id}/status
to get the current status (and trigger a WS notification if it changed).

Desktop mode: No Flow entity, single @local compute node, simplified execution.
"""

from __future__ import annotations

import logging
from typing import Any

from flow_sdk.api.api_types.api_field import APIField
from flow_sdk.core import Entity, action
from flow_sdk.fs_records.agent_status import is_running as _is_running_status
from flow_sdk.request_context.methods import get_current_request_info
from flow_sdk.responses.response import ApiFailResponse, ApiSuccessResponse

from flow_sdk.builtin.agentic_process.agentic_process import AgenticProcess

logger = logging.getLogger(__name__)


class AgenticProcessor(Entity):
    _api_visible = True
    type: str = APIField(default="agentic_processor")

    worker_type: str = APIField(default="claude")
    active_agentic_process_id: str | None = APIField(default=None)

    async def _is_running(self) -> bool:
        """True when the active process has a running transcript-derived status."""
        if not self.active_agentic_process_id:
            return False
        proc = await AgenticProcess.get_by_id(self.active_agentic_process_id)
        return bool(proc and _is_running_status(proc.status))

    @action.all(action_name="runFile")
    async def run_file(self, vfs_path: str | None = None):
        """Run an instruction file from VFS path.

        Loads the file content and creates an AgenticProcess for execution.

        Args:
            vfs_path: VFS path to the instruction file

        Returns:
            AgenticProcess entity data
        """
        if await self._is_running():
            return ApiFailResponse(message="Processor is already running")

        if not vfs_path:
            return ApiFailResponse(message="vfs_path is required")

        try:
            request_info = get_current_request_info()
            owner = request_info.someone_typeid if request_info else None

            file_content = None
            try:
                from flow_sdk.api.fs_api import VFSPath
                from pathlib import Path

                vfs = VFSPath(vfs_path)
                local_path = vfs.local_path if hasattr(vfs, "local_path") else None
                if local_path and Path(local_path).exists():
                    file_content = Path(local_path).read_text()
            except Exception as e:
                logger.warning(f"Could not load file from VFS path {vfs_path}: {e}")

            process = AgenticProcess(
                processor_id=self.id,
                instruction_content=file_content or "",
                source_vfs_path=vfs_path,
            )
            await process.save(owner)

            self.active_agentic_process_id = process.id
            await self.save(owner)

            return ApiSuccessResponse(
                data={
                    "id": process.id,
                    "type": process.type,
                    "processor_id": process.processor_id,
                    "instruction_content": process.instruction_content,
                    "source_vfs_path": process.source_vfs_path,
                }
            )

        except FileNotFoundError:
            logger.error(f"AgenticProcessor {self.id} file not found: {vfs_path}")
            return ApiFailResponse(message=f"File not found: {vfs_path}")

        except Exception as e:
            logger.exception(f"AgenticProcessor {self.id} runFile error: {e}")
            return ApiFailResponse(message=str(e))

    @action.all(action_name="run")
    async def run(
        self,
        instruction_content: str | None = None,
        context: dict[str, Any] | None = None,
    ):
        """Run instruction content with context — returns AgenticProcess.

        Creates an AgenticProcess and returns its entity data. The process
        is then driven via the process.open() / process.prompt() API.

        Args:
            instruction_content: The instruction content to execute
            context: Context data (workdir, env_vars, model, etc.)

        Returns:
            AgenticProcess data (id, type, etc.)
        """
        if await self._is_running():
            return ApiFailResponse(message="Processor is already running")

        if not instruction_content:
            return ApiFailResponse(message="instruction_content is required")

        logger.info(f"AgenticProcessor {self.id}: run started, content_len={len(instruction_content)}")
        try:
            request_info = get_current_request_info()
            owner = request_info.someone_typeid if request_info else None

            process = AgenticProcess(
                processor_id=self.id,
                instruction_content=instruction_content,
                context_data=context or {},
            )
            await process.save(owner)
            logger.info(f"AgenticProcessor {self.id}: created process {process.id}")

            self.active_agentic_process_id = process.id
            await self.save(owner)

            return ApiSuccessResponse(
                data={
                    "id": process.id,
                    "type": process.type,
                    "processor_id": process.processor_id,
                    "instruction_content": process.instruction_content,
                    "context": process.context_data,
                }
            )

        except Exception as e:
            logger.exception(f"AgenticProcessor {self.id} run error: {e}")
            return ApiFailResponse(message=str(e))

    @action.all(action_name="execute")
    async def execute(
        self,
        instruction_content: str | None = None,
        context: dict[str, Any] | None = None,
    ):
        """Execute instruction content directly — returns AgenticProcess.

        Simpler API than run(): takes instruction text and context directly.

        Args:
            instruction_content: Plain text instruction content
            context: Context data (workdir, env_vars, model, etc.)

        Returns:
            AgenticProcess data (id, type, etc.)
        """
        if await self._is_running():
            return ApiFailResponse(message="Processor is already running")

        if not instruction_content:
            return ApiFailResponse(message="instruction_content is required")

        logger.info(f"AgenticProcessor {self.id}: execute started, content_len={len(instruction_content)}")
        try:
            request_info = get_current_request_info()
            owner = request_info.someone_typeid if request_info else None

            process = AgenticProcess(
                processor_id=self.id,
                instruction_content=instruction_content,
                context_data=context or {},
            )
            await process.save(owner)
            logger.info(f"AgenticProcessor {self.id}: created process {process.id}")

            self.active_agentic_process_id = process.id
            await self.save(owner)

            return ApiSuccessResponse(
                data={
                    "id": process.id,
                    "type": process.type,
                    "processor_id": process.processor_id,
                    "instruction_content": process.instruction_content,
                    "context": process.context_data,
                }
            )

        except Exception as e:
            logger.exception(f"AgenticProcessor {self.id} execute error: {e}")
            return ApiFailResponse(message=str(e))

    @action.all(action_name="createProcess")
    async def create_process(
        self,
        context: dict[str, Any] | None = None,
        result: dict[str, Any] | None = None,
        visible: bool = False,
    ):
        """Create a new idle process ready for open() / prompt() calls.

        Args:
            context: Context data (workdir, env_vars, model, etc.)
            result: Optional ProcessResult metadata

        Returns:
            AgenticProcess entity data in IDLE status
        """
        try:
            request_info = get_current_request_info()
            owner = request_info.someone_typeid if request_info else None

            from flow_sdk.builtin.cli_workers.claude_cli import ClaudeCliOptions

            context_data = dict(context or {})
            workdir = context_data.pop("workdir", None)

            fork_session = bool(context_data.pop("fork_session", False))
            resume_session_id = context_data.pop("resume_session_id", None)

            additional_dirs: list[str] = list(context_data.pop("additional_dirs", None) or [])

            cli_opts = ClaudeCliOptions(
                model=context_data.pop("model", None) or None,
                permission_mode=context_data.pop("permission_mode", "bypassPermissions"),
                chrome=bool(context_data.pop("chrome", False)),
                debug=bool(context_data.pop("debug", True)),
                worktree=bool(context_data.pop("worktree", False)),
                agents_json=context_data.pop("agents_json", None),
            )

            if fork_session and resume_session_id:
                cli_opts.resume = True
                cli_opts.fork_session_id = resume_session_id
            elif resume_session_id:
                cli_opts.resume = True

            process = AgenticProcess(
                processor_id=self.id,
                instruction_content="",
                cli_config=cli_opts.to_json(),
                context_data=context_data,
                workdir=workdir,
                visible=visible,
                additional_dirs=additional_dirs,
            )
            if resume_session_id and not fork_session:
                process.worker_session_id = resume_session_id
            await process.save(owner)

            if result and isinstance(result, dict):
                try:
                    from flow_sdk.builtin.process_result import ProcessResult

                    result_uname = result.get("uname")
                    existing_result = None
                    if result_uname:
                        existing_result = await ProcessResult.get_by_uname(result_uname)

                    if existing_result:
                        existing_result.agentic_process_id = process.id
                        existing_result.status = "running"
                        existing_result.result_type = result.get("result_type")
                        existing_result.source_session_id = result.get("source_session_id")
                        await existing_result.save(owner)
                    else:
                        process_result = ProcessResult(
                            uname=result_uname,
                            agentic_process_id=process.id,
                            status="running",
                            result_type=result.get("result_type"),
                            source_session_id=result.get("source_session_id"),
                        )
                        await process_result.save(owner)
                except ImportError:
                    logger.debug("ProcessResult entity not available, skipping result creation")

            self.active_agentic_process_id = process.id
            await self.save(owner)

            logger.info(f"AgenticProcessor {self.id} created process {process.id}")

            return ApiSuccessResponse(
                data={
                    "id": process.id,
                    "type": process.type,
                    "processor_id": process.processor_id,
                }
            )

        except Exception as e:
            logger.exception(f"AgenticProcessor {self.id} createProcess error: {e}")
            return ApiFailResponse(message=str(e))
